### PR TITLE
feat: create a simple inbound socket implementation

### PR DESCRIPTION
**Pull request by marco-ippolito**
I woud like to start collaboration on what a possible server socket implementation could look like.
Its working but its ugly and leaking 😄 but a first step

### CODE REVIEW

**Comment by mcollina** [comment]
I think you need an abstraction that would manage HTTP/1, HTTP/2 and one day HTTP/3

**Review by sheplu — APPROVED** [review]
LGTM :D

**Comment by marco-ippolito** [comment]
> I think you need an abstraction that would manage HTTP/1, HTTP/2 and one day HTTP/3

I was thinking that the inbound tpc socket is the lowest level abstraction, like udp socket for http3 and create a layer ontop that abstract the differences

**Comment by mcollina** [comment]
I would recommend not to use EventEmitter for the lower level, as we need the best possible speed.

Check out the work that was done for bob:

https://www.npmjs.com/package/bob-streams

**Comment by wesleytodd** [comment]
I agree with @mcollina on this, we should make these lower level api's as performant as possible even if it "breaks user expectations", if we wanted to wrap it in an EE for users who do want to use the api directly, we can later.

File: packages/milo-http-server-addon/lib/http-server.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+				throw new Error("HTTP/3 is not supported yet");
+			default:
+				throw new Error("Invalid HTTP version");
+		}
+	}
+
+	listen(cb) {
+		const parsing = (req, res, err) => {
+			const parser = load();
+			const data = Buffer.from(req.data);
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/lib/http-server.js]: here we should actually return `req.data` as buffer not a string

File: packages/milo-http-server-addon/lib/socket.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import bindings from "bindings";
+const { InboundTCPSocket } = bindings("inbound-tcp-socket");
+
+export class InboundSocket {
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/lib/socket.js]: probably we should rename it to `InboundTCPSocket`

File: packages/milo-http-server-addon/lib/response.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+export class Response {
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/lib/response.js]: We might want to use undici implementation if available

File: packages/milo-http-server-addon/lib/request.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export class Request {
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/lib/request.js]: same here we should use undici implementation

File: packages/milo-http-server-addon/src/inbound_tcp_socket.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+void after_write(uv_write_t *req, int status)
+{
+    free(req);
+}
+
+static Napi::Value write(const Napi::CallbackInfo &info)
+{
+    uv_stream_t *client = (uv_stream_t *)info.Data();
+    Napi::String data = info[0].As<Napi::String>();
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/src/inbound_tcp_socket.cpp]: data should be a buffer not a string

File: packages/milo-http-server-addon/src/inbound_tcp_socket.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    if (nread <= 0)
+    {
+        free(buf->base);
+        return;
+    }
+    server_data *client_data = (server_data *)(client->data);
+    Napi::Env env = client_data->callback.Env();
+    Napi::HandleScope scope(env);
+    Napi::Object request = Napi::Object::New(env);
+    request.Set("data", Napi::String::New(env, buf->base, nread));
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/src/inbound_tcp_socket.cpp]: data should be a buffer not a string

File: packages/milo-http-server-addon/test/http-server.test.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import { HttpServer } from '../lib/http-server.js';
+import { test } from 'node:test';
+import assert from 'node:assert';
+import { Response } from '../lib/response.js';
+
+test('HttpServer', async () => {
+    const body = "Hello, World!\r\n";
+    const server = new HttpServer({ hostname: "0.0.0.0", port: 8001 });
+    server.listen((req, err) => {
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/test/http-server.test.js]: currently errors are swallowed for some reason 😅

File: packages/milo-http-server-addon/src/inbound_tcp_socket.cpp [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+    int result = 0;
+    result = uv_tcp_init(eventLoop, &this->tcpServer);
+    if (result != 0)
+    {
+        emit_uv_error(callback, result);
+        return;
+    }
+
+    struct sockaddr_in address;
```
Comment (marco-ippolito) [inline packages/milo-http-server-addon/src/inbound_tcp_socket.cpp]: we can delegate this part to the constructor

File: packages/milo-http-server-addon/lib/response.js (hunk 3 above)
Comment (mcollina) [reply packages/milo-http-server-addon/lib/response.js]: you can import undici

File: packages/milo-http-server-addon/lib/request.js (hunk 4 above)
Comment (mcollina) [reply packages/milo-http-server-addon/lib/request.js]: import if from undici

**Comment by wesleytodd** [comment]
Can we merge this? I would like to make this so we can start opening more code PRs and duplicating some of the basic setup this PR has seems unreliable.

**Pull request merged**